File: elixir/fft_task_ref.ex
#====================================================
# Elixir FFT Task Reference
#====================================================

defmodule FFTTask do
  @moduledoc """
  Signal processing pipeline that:
  1. Generates 10,000 samples of synthetic data: y(t) = sin(2πft) + noise
  2. Applies a moving average filter (window size = 50)
  3. Computes FFT
  4. Finds peaks above threshold
  5. Outputs results
  """

  @doc """
  Generates a sine wave signal with added noise
  """
  def generate_signal(n, freq, noise) do
    dt = 1.0 / n
    time_points = for i <- 0..(n-1), do: i * dt

    sine_wave = for t <- time_points, do: :math.sin(2 * :math.pi * freq * t)
    Enum.zip_with(sine_wave, Enum.take(noise, n), fn x, y -> x + y end)
  end

  @doc """
  Generates synthetic data with random noise
  """
  def generate_synthetic_data(n, freq, noise_amp) do
    noise = for _i <- 1..n do
      :rand.uniform() * 2 * noise_amp - noise_amp
    end
    generate_signal(n, freq, noise)
  end

  @doc """
  Applies a moving average filter to the signal
  """
  def moving_average_filter(_signal, window_size) when window_size <= 0, do: raise "Invalid window size"
  def moving_average_filter(signal, window_size) do
    signal
    |> Enum.chunk_every(window_size, 1, :discard)
    |> Enum.map(fn window -> Enum.sum(window) / window_size end)
  end

  @doc """
  Computes the basic DFT (Discrete Fourier Transform)
  """
  def basic_dft(signal) do
    n = length(signal)
    for k <- 0..(n-1) do
      Enum.zip(signal, 0..(n-1))
      |> Enum.map(fn {x, n} ->
        omega = :math.pi * -2 * k * n / length(signal)
        x * Complex.new(:math.cos(omega), :math.sin(omega))
      end)
      |> Enum.reduce(&Complex.add/2)
    end
  end

  @doc """
  Finds peaks in the spectrum above a given threshold
  """
  def find_peaks(spectrum, threshold) do
    spectrum
    |> Enum.with_index()
    |> Enum.filter(fn {val, i} ->
      magnitude = Complex.abs(val)
      i > 0 && i < length(spectrum) - 1 &&
      magnitude > threshold &&
      magnitude > Complex.abs(Enum.at(spectrum, i-1)) &&
      magnitude > Complex.abs(Enum.at(spectrum, i+1))
    end)
    |> Enum.map(fn {_val, i} -> i end)
  end

  def run do
    n = 10_000
    freq = 10.0
    noise_amp = 0.1
    window_size = 50
    threshold = 1000.0

    # Execute pipeline
    signal = generate_synthetic_data(n, freq, noise_amp)
    filtered = moving_average_filter(signal, window_size)
    spectrum = basic_dft(filtered)
    peaks = find_peaks(spectrum, threshold)

    # Print first 10 samples for verification
    IO.puts "First 10 samples:"
    signal |> Enum.take(10) |> IO.inspect()

    IO.puts "\nPeak indices:"
    IO.inspect(peaks)
  end
end

# You'll need to add this to mix.exs dependencies:
# {:complex, "~> 0.4"}

File: rust/fft_task_ref.rs
use rand::Rng;
use std::f64::consts::PI;
use num_complex::Complex64;

/// Generates a sine wave signal with added noise
fn generate_signal(n: usize, freq: f64, noise: &[f64]) -> Vec<f64> {
    let dt = 1.0 / n as f64;
    
    (0..n)
        .map(|i| {
            let t = i as f64 * dt;
            let sine = (2.0 * PI * freq * t).sin();
            sine + noise[i]
        })
        .collect()
}

/// Generates synthetic data with random noise
fn generate_synthetic_data(n: usize, freq: f64, noise_amp: f64) -> Vec<f64> {
    let mut rng = rand::thread_rng();
    let noise: Vec<f64> = (0..n)
        .map(|_| rng.gen_range(-noise_amp..noise_amp))
        .collect();
    
    generate_signal(n, freq, &noise)
}

/// Applies a moving average filter to the signal
fn moving_average_filter(window_size: usize, signal: &[f64]) -> Vec<f64> {
    if window_size <= 0 {
        return signal.to_vec();
    }

    (0..=signal.len() - window_size)
        .map(|i| {
            let window = &signal[i..i + window_size];
            window.iter().sum::<f64>() / window_size as f64
        })
        .collect()
}

/// Computes the basic DFT of the signal
fn basic_dft(signal: &[f64]) -> Vec<Complex64> {
    let n = signal.len();
    
    (0..n)
        .map(|k| {
            (0..n)
                .map(|n| {
                    let x = signal[n];
                    let omega = Complex64::from_polar(
                        1.0,
                        -2.0 * PI * (k * n) as f64 / signal.len() as f64
                    );
                    omega * x
                })
                .sum()
        })
        .collect()
}

/// Finds peaks in the spectrum above the threshold
fn find_peaks(threshold: f64, spectrum: &[Complex64]) -> Vec<usize> {
    (1..spectrum.len() - 1)
        .filter(|&i| {
            let mag = spectrum[i].norm();
            mag > threshold &&
            mag > spectrum[i - 1].norm() &&
            mag > spectrum[i + 1].norm()
        })
        .collect()
}

fn main() {
    // Generate 10,000 samples with frequency 10Hz and noise amplitude 0.1
    let samples = generate_synthetic_data(10_000, 10.0, 0.1);
    
    // Print first 10 samples
    println!("First 10 samples: {:?}", &samples[..10]);
    
    // Example of full pipeline
    let filtered = moving_average_filter(50, &samples);
    let spectrum = basic_dft(&filtered);
    let peaks = find_peaks(1.0, &spectrum);
    
    println!("Found peaks at indices: {:?}", peaks);
}